src/bsoncxx/exception/error_category.cpp

```diff
@@ -0,0 +1,110 @@
+// Copyright 2015 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <bsoncxx/config/prelude.hpp>
+
+#include <system_error>
+
+#include <bsoncxx/exception/error_category.hpp>
+
+#include <bsoncxx/exception/bsoncxx_error.hpp>
+#include <bsoncxx/exception/libbson_error.hpp>
+
+namespace {
+
+using namespace bsoncxx;
+
+///
+/// An error category for errors originating in bsoncxx.
+///
+class bsoncxx_error_category_impl final : public std::error_category {
```

Comment:
It shouldn't matter because they are still in an anonymous namespace, so they won't be leaked out anywhere.  To a caller, the result of libbson_error_category() or bsoncxx_error_category() will just look like any error_category.Though for consistency's sake maybe we should pick one approach?

I split it up like this:
- bsoncxx_error.hpp/cpp - defines driver-specific error types and transforms them into error_codes
- libbson_error.hpp/cpp - ditto above, for libbson
- error_category.hpp/cpp - defines both error categories

It looks like you/Matt split it up like this:
- mongoc_error.hpp - defines libmongoc-specific error types
- mongocxx_error.hpp - ditto above, for mongocxx
- mongoc_error_category.hpp/cpp - defines mongoc error category, and transforms mongoc_error types into error_codes
- mongocxx_error_category.hpp/cpp - ditto above, for mongocxx

Personally, I think it makes more sense to keep the utilities (like make_error_code) with the error type files rather than with the error_categories.  But they are all related to each other so you could argue either way!  Let's pick one.